candies/cli/parsers/standard.py

```diff
@@ -24,33 +25,77 @@ def concat(left: Arg['left operand'], right: Arg['right operand']):
     The parser may also be provided with a help text (taken from the function's
     docstring) and default values for arguments.
 
-    Args:
-        func (function): A function to construct an instance of
-            `ArgumentParser` from.
+    Attributes:
+        cli: An instance of `CLI` to initialise a parser for.
     """
 
-    def __init__(self, func: Callable):
-        self._func = func
-
-    def parse(self, args: List[str]) -> Dict[str, Any]:
-        signature = inspect.signature(self._func)
+    def __init__(self, cli):
+        self.cli = cli
 
+    def parse(self, args: Iterable[str]) -> Iterable[Invocation]:
         parser = ArgumentParser()
-        parser.description = self._func.__doc__
 
-        # Add parameters to the parser.
-        for parameter in signature.parameters.values():
-            parser.add_argument(f'-{parameter.name[0]}',
-                                f'--{parameter.name}',
-                                **dict(configuration(parameter)))
+        configure(parser, self.cli.main)
 
-        args = parser.parse_args(args)
+        args = parser.parse_args(list(args))
         args = args.__dict__
 
-        return args
+        # It's assumed that the entries preserve the insertion order.
+        # This feature was introduced in Python 3.7.
+        # It's required to be able to differentiate arguments for each command.
+        invocations = [Invocation(args={})]
+
+        for key, value in args.items():
+            # Read arguments until we find '.command', '..command', etc.
+            if key.startswith('.'):
+                # `value` may be None in a case
+                # if a subcommand was not specified.
+                if value is None:
+                    break
+
+                invocations[-1].next = value
+                invocations.append(Invocation(args={}))
+            else:
+                invocations[-1].args[key] = value
+
+        return invocations
+
+
+def configure(parser, command, prefix='.'):
```

Comment:
Shouldn't we use type annotations here as well?

Reply:
Yes, we should. I removed them because I had a circular import. However, this is not the case anymore, so I'll add them back.